File: src/main/java/io/github/takenoha/towerdefense/paper/ResearchCrystalItemIdentity.java
package io.github.takenoha.towerdefense.paper;

import java.util.Objects;
import java.util.UUID;

/** Team and issuance-batch identity carried by a delivered research crystal. */
public record ResearchCrystalItemIdentity(
        UUID batchId,
        UUID teamId,
        int issuedQuantity,
        Integer segmentOffset,
        Integer segmentQuantity) {
    public ResearchCrystalItemIdentity(UUID batchId, UUID teamId, int issuedQuantity) {
        this(batchId, teamId, issuedQuantity, null, null);
    }

    public ResearchCrystalItemIdentity {
        Objects.requireNonNull(batchId, "batchId");
        Objects.requireNonNull(teamId, "teamId");
        if (issuedQuantity <= 0) {
            throw new IllegalArgumentException("issuedQuantity must be positive");
        }
        if ((segmentOffset == null) != (segmentQuantity == null)) {
            throw new IllegalArgumentException(
                    "segmentOffset and segmentQuantity must be supplied together");
        }
        if (segmentOffset != null
                && (segmentOffset < 0
                        || segmentQuantity <= 0
                        || (long) segmentOffset + segmentQuantity > issuedQuantity)) {
            throw new IllegalArgumentException("research crystal segment is outside the batch");
        }
    }

    public boolean hasSegmentIdentity() {
        return segmentOffset != null;
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/paper/ResearchCrystalItemIdentity.kt
package io.github.takenoha.towerdefense.paper

import java.util.Objects
import java.util.UUID
import kotlin.jvm.JvmRecord

/** Team and issuance-batch identity carried by a delivered research crystal. */
@JvmRecord
data class ResearchCrystalItemIdentity(
    val batchId: UUID,
    val teamId: UUID,
    val issuedQuantity: Int,
    val segmentOffset: Int?,
    val segmentQuantity: Int?,
) {
    constructor(batchId: UUID, teamId: UUID, issuedQuantity: Int) :
        this(batchId, teamId, issuedQuantity, null, null)

    init {
        Objects.requireNonNull(batchId, "batchId")
        Objects.requireNonNull(teamId, "teamId")
        require(issuedQuantity > 0) { "issuedQuantity must be positive" }
        require((segmentOffset == null) == (segmentQuantity == null)) {
            "segmentOffset and segmentQuantity must be supplied together"
        }
        if (segmentOffset != null &&
            (segmentOffset < 0 || segmentQuantity!! <= 0 ||
                segmentOffset.toLong() + segmentQuantity > issuedQuantity)
        ) {
            throw IllegalArgumentException("research crystal segment is outside the batch")
        }
    }

    fun hasSegmentIdentity(): Boolean = segmentOffset != null
}
